polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -0,0 +1,147 @@
+//===- LoopInternalization.cpp - Promote memory access to local memory ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass tiles perfect loop nests to 'prefetch' memory accesses in shared
+// local memory.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/LoopUtils.h"
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/IR/FunctionInterfaces.h"
+#include "mlir/Interfaces/LoopLikeInterface.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "loop-internalization"
+
+namespace mlir {
+namespace polygeist {
+#define GEN_PASS_DEF_LOOPINTERNALIZATION
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h.inc"
+} // namespace polygeist
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+/// Collect perfectly nested loops starting from \p root.  Loops are
+/// perfectly nested if each loop is the first and only non-terminator operation
+/// in the parent loop.
+template <typename T, typename = std::enable_if_t<llvm::is_one_of<
+                          T, affine::AffineForOp, scf::ForOp>::value>>
+void getPerfectlyNestedLoops(SmallVector<T> &nestedLoops, T root) {
+  for (unsigned i = 0; i < std::numeric_limits<unsigned>::max(); ++i) {
+    nestedLoops.push_back(root);
+    Block &body = root.getLoopBody().front();
```

Comment:
How would this detect a case where the parent loop contains more than one block?

Reply:
If that is already checked by the verifiers of `affine.for` and `scf.for`, we don't need to assert that here.

Reply:
Looks like is not explicitly checked in the verifiers, but seems the builder can only build one block. We can add assertion anyways in case one may want to extend the template to allow more loop operations.